### PR TITLE
no need to use region peers

### PR DESCRIPTION
**Pull request by Connor1996**
fix performance issue

### CODE REVIEW

**Comment by Connor1996** [comment]
/run-integration-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by Connor1996** [comment]
/run-integration-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by Connor1996** [comment]
/run-integration-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by siddontang** [comment]
em, don't we need this in master, only in release?

**Comment by Connor1996** [comment]
no need in master cause master branch is not implemented in that way due to threaded raftstore

**Comment by Connor1996** [comment]
/run-integration-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by Connor1996** [comment]
PTAL @BusyJay

**Comment by Connor1996** [comment]
/run-integration-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -966,15 +965,25 @@ impl Peer {
...
+            // but it is okay cause this case would never happen.
+            // The most possible case is like this, but is abandoned in propose stage.
+            //     | region 1 |       region 2      |  store 1
+            //                   || split 2 into 3
+            //                   \/
+            //     | region 1 | region 2 | region 3 |
+            //                   || merge 2 into 1 (never happen, cause source region's log gap should not contain AdminRequest(split 2 into 3))
+            //                   \/
+            //     |        region 1     | region 3 |  store 2,3
+            if let Some((_, r)) = region_ranges
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: do we need to add some tests here?

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -966,15 +965,25 @@ impl Peer {
...
+                .range((Unbounded::<Vec<u8>>, Included(enc_end_key(&region))))
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Is it a valid range?

File: src/raftstore/store/peer.rs (hunk 1 above)
Comment (Connor1996) [reply src/raftstore/store/peer.rs]: already covered by merge tests

File: src/raftstore/store/peer.rs (hunk 2 above)
Comment (Connor1996) [reply src/raftstore/store/peer.rs]: What do you mean?

**Comment by hicqu** [comment]
LGTM.

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -966,15 +965,25 @@ impl Peer {
...
-            if let Some(r) = region_ranges
-                .range((Excluded(enc_start_key(&region)), Unbounded::<Vec<u8>>))
+            // Note that: below logic can not find the overlap region with range (a, b] that a < snap_region.end_key < b,
+            // but it is okay cause this case would never happen.
+            // The most possible case is like this, but is abandoned in propose stage.
+            //     | region 1 |       region 2      |  store 1
+            //                   || split 2 into 3
+            //                   \/
+            //     | region 1 | region 2 | region 3 |
+            //                   || merge 2 into 1 (never happen, cause source region's log gap should not contain AdminRequest(split 2 into 3))
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: What if merging `region 1` into `region 2`?

Comment (Connor1996) [reply src/raftstore/store/peer.rs]: `region 1` update merge target, and then `region 2` receives a new snapshot and decide to destroy `region 1`. And here logic can detect `region 1`'s range is overlapped with `region 2`'s snapshot, so wait here until the destroy of `region 1` finish.

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -966,15 +965,25 @@ impl Peer {
             let region = snap_data.take_region();
             // For merge process, when applying snapshot or create new peer the stale source peer is destroyed asynchronously.
             // So here checks whether there is any overlap, if so, wait and do not handle raft ready.
-            if let Some(r) = region_ranges
-                .range((Excluded(enc_start_key(&region)), Unbounded::<Vec<u8>>))
+            // Note that: below logic can not find the overlap region with range (a, b] that a < snap_region.end_key < b,
+            // but it is okay cause this case would never happen.
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Better add some `assert`s to ensure that even it happens due to bugs, data will not get corrupted.

**Comment by Connor1996** [comment]
/run-all-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by zhouqiang-cl** [comment]
/run-all-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by zhouqiang-cl** [comment]
/run-all-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by zhouqiang-cl** [comment]
/run-integration-ddl-test tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by Connor1996** [comment]
/run-integration-ddl-test tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Comment by zhouqiang-cl** [comment]
/run-integration-ddl-test tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Pull request merged**